### DISCUSSION TITLE
How to print PRF values for a NER model using spaCy v3?

### DISCUSSION THREAD

**Discussion by CarolLuca** [explosion/spaCy#8178] [Help: Coding & Implementations] [answered]
# Context
I am currently working on a NER model for the Romanian Legal Domain. I began creating a custom model using spaCy v2 (v2.2.4), for which I successfully implemented a code to find the PRF values. Now, after I made the transition to spaCy v3 (v3.0.6), I find it difficult to evaluate the performance of my model.
# Problem
I tried to do the following:
* Use the same code in spaCy v3.0.6. like that for spaCy v2.2.4 (problem: GoldParser is not present in spaCy v3.0.6)
* Use spaCy v2.2.4 to train the v3.0.6 model (problem: I think that the models are not saved in the same way regardless of their version)
* Use get_ner_prf() (problem: I did not understand how to create the parameter of type Example and I am also not sure how to call the function)
# Resources
Here is a list of all the resources I have at the moment:
* Config files for the v3.0.6 model (and all the other necessary files)
* Train and test data in the old spaCy format
* Saved v3.0.6 custom model for Romanian
# Request
I would be grateful to receive a code that works for spaCy v3.0.6 and calculates the PRF values - preferably individual results for every entity type. Also, it will be great if the code only makes use of the resources aforementioned. If any other information is needed, I am glad to send it.

**Comment by svlandeg** [comment, marked as answer]
Hi!

Unfortunately a model built with spaCy v2 will be incompatible with v3. So you'll either have to retrain the model, or stick to v2 if you want to keep using your old model.

The [migration guide](https://spacy.io/usage/v3#migrating) contains pointers on some of the things that have changed in v3. `GoldParse` has indeed been replaced with `Example` - you can find a little bit more information on that [here](https://spacy.io/api/data-formats#dict-input). In short, an `Example` object contains two `Doc` objects, one holding gold-standard reference annotations (`reference`) and one holding the predictions of your pipeline (`predicted`). So one option is to create the gold-standard `Doc` objects in code, and use those to make a new `Example` object with:
```
reference = Doc(...)
reference.ents = ...
predicted = nlp(reference.text)
example = Example(predicted, reference)
```

If you want to convert your old data to the binary format `.spacy`, you can also have a look here: https://spacy.io/api/cli#convert. The `.spacy` format contains `DocBin` objects which is just a container of `Doc` objects:
```
doc_bin = DocBin().from_disk("./data.spacy")
docs = list(doc_bin.get_docs(nlp.vocab))
```

You can use spaCy's built-in reader to yield `Example` objects from this format directly:
```
from spacy.training import Corpus

corpus = Corpus("./data.spacy")
train_data = corpus(nlp)
```
And once you have these `Example` objects, you should be able to call `nlp.evaluate(examples)` or `scorer.get_ner_prf(examples)` directly.

Hope this helps!

**Reply by CarolLuca** [reply]
Hi!

Thank you very much for the help! 
The given explanations were very useful and together with the discussion at the following link:
https://stackoverflow.com/questions/67229277/spacy-3-ner-scorer-throws-typeerror-score-takes-2-positional-arguments-but
I was able to write the following code (in case anyone that reads the current discussion may still find it difficult to make the transition):
```python
# TRAIN_DATA format
TRAIN_DATA = [
    ("Who was Hamlet?", [(8,14,'PER')]),
    ("Have you ever met Rome?", [(18,22,'LOC')])
]

import spacy 
from spacy.tokens import Span
from spacy import displacy
from spacy.training import *
from spacy.scorer import Scorer
from spacy.util import minibatch, compounding

nlp = spacy.load("./output/model-best")

# evaluate function
def evaluate(ner_model, testing_data):
    scorer = Scorer()
    examples = []
    for input_, annot in testing_data:
        doc_gold_text = ner_model.make_doc(input_)
        example = Example.from_dict(doc_gold_text, {"entities": annot})
        example.predicted = ner_model(input_)
        examples.append(example)
        
    return scorer.score(examples)

# print the results
print(evaluate(nlp, TRAIN_DATA))
```
The output is in the following format:
```python
{'token_acc': 1.0, 'token_p': 1.0, 'token_r': 1.0, 'token_f': 1.0, 'sents_p': 1.0, 'sents_r': 1.0, 'sents_f': 1.0, 'tag_acc': None, 'pos_acc': None, 'morph_acc': None, 'morph_per_feat': None, 'dep_uas': None, 'dep_las': None, 'dep_las_per_type': None, 'ents_p': 0.9833333333333333, 'ents_r': 0.9365079365079365, 'ents_f': 0.9593495934959351, 'ents_per_type': {'LAW': {'p': 1.0, 'r': 1.0, 'f': 1.0}, 'LOC': {'p': 1.0, 'r': 0.6666666666666666, 'f': 0.8}, 'ORG': {'p': 0.9565217391304348, 'r': 0.9565217391304348, 'f': 0.9565217391304348}, 'TIME': {'p': 1.0, 'r': 1.0, 'f': 1.0}, 'PER': {'p': 1.0, 'r': 1.0, 'f': 1.0}}, 'cats_score': 0.0, 'cats_score_desc': 'macro F', 'cats_micro_p': 0.0, 'cats_micro_r': 0.0, 'cats_micro_f': 0.0, 'cats_macro_p': 0.0, 'cats_macro_r': 0.0, 'cats_macro_f': 0.0, 'cats_macro_auc': 0.0, 'cats_f_per_type': {}, 'cats_auc_per_type': {}}
```
If anything seems wrong, please don't hesitate to mention.
Thanks again!

**Reply by svlandeg** [reply]
Happy to hear you got it working, and thanks for sharing the full solution!
